### PR TITLE
Create a data model for a domain application

### LINKED ISSUE #221: Define a basic domain application model
_Outcome_: Add an Application model migration to our schema and define its basic attributes.

### PR DESCRIPTION
**Pull request by neilmb**
# Domain application data model

## 🗣 Description ##

This creates a table (or three) for storing data from a new domain application. The fields here are derived from the Figma prototype of what information we will collect on that application form. Contact information and domain names appear on that form where applicants can add unlimited numbers so they get new models and ManyToMany relations to the DomainApplication.

## 💭 Motivation and context ##

Closes #221

### CODE REVIEW

File: src/registrar/models/models.py [hunk 1]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+    STATUS_CHOICES = [
+        (STARTED, STARTED),
+        (SUBMITTED, SUBMITTED),
+        (INVESTIGATING, INVESTIGATING),
+        (APPROVED, APPROVED),
+    ]
+    status = FSMField(
+        choices=STATUS_CHOICES,  # possible states as an array of constants
+        default=STARTED,  # sensible default
+        protected=True,  # cannot change state directly, must use methods!
```
Comment (SSPJ) [inline src/registrar/models/models.py:143]: I've been using this branch for my audit logging work and I've noticed that you cannot make changes within the Django admin with this set to True.

I'm not sure what to advise. I think we can defer any work on the admin for right now, but perhaps setting this to `False` would make future dev work a little easier? OTOH, I like that it is catching instances of changing state without using the proper methods.

File: src/registrar/models/models.py [hunk 2]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+        (COUNTY, "a county, parish, or borough"),
+        (CITY, "a city, town, township, village, etc."),
+        (SPECIAL_DISTRICT, "an independent organization within a single state"),
+    ]
+    organization_type = models.CharField(
+        max_length=255, choices=ORGANIZATION_CHOICES, help_text="Type of Organization"
+    )
+
+    EXECUTIVE = "Executive"
+    JUDICIAL = "Judicial"
```
Comment (SSPJ) [inline src/registrar/models/models.py:186]: **Suggestion:** I kinda dislike having constants mixed with field declarations. If you like it better this way, I'm fine with it, but if you don't care, let's put constants at the top.

File: src/registrar/models/models.py [hunk 3]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+    )
+
+    organization_name = models.TextField(null=True, help_text="Organization name")
+    street_address = models.TextField(null=True, help_text="Street Address")
+    unit_type = models.CharField(max_length=15, null=True, help_text="Unit type")
+    unit_number = models.CharField(max_length=255, null=True, help_text="Unit number")
+    state_territory = models.CharField(
+        max_length=2, null=True, help_text="State/Territory"
+    )
+    zip_code = models.CharField(max_length=10, null=True, help_text="ZIP code")
```
Comment (SSPJ) [inline src/registrar/models/models.py:207]: These need to be indexed.

File: src/registrar/models/models.py [hunk 4]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+        choices=BRANCH_CHOICES,
+        null=True,
+        help_text="Branch of federal government",
+    )
+
+    is_election_office = models.BooleanField(
+        null=True, help_text="Is your ogranization an election office?"
+    )
+
+    organization_name = models.TextField(null=True, help_text="Organization name")
```
Comment (SSPJ) [inline src/registrar/models/models.py:200]: This needs to be indexed.

File: src/registrar/models/models.py [hunk 5]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+    zip_code = models.CharField(max_length=10, null=True, help_text="ZIP code")
+
+    authorizing_official = models.ForeignKey(
+        Contact,
+        null=True,
+        related_name="authorizing_official",
+        on_delete=models.PROTECT,
+    )
+
+    # "+" means no reverse relation to lookup applications from Website
```
Comment (SSPJ) [inline src/registrar/models/models.py:216]: Kudos for relevant and timely code comment. 💯

File: src/registrar/models/models.py [hunk 6]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+        on_delete=models.PROTECT,
+    )
+    alternative_domains = models.ManyToManyField(Website, related_name="alternatives+")
+
+    submitter = models.ForeignKey(
+        Contact,
+        null=True,
+        related_name="submitted_applications",
+        on_delete=models.PROTECT,
+    )
```
Comment (SSPJ) [inline src/registrar/models/models.py:233]: `submitter` is `Contact` not `User`?

File: src/registrar/models/models.py [hunk 7]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+
+    acknowledged_policy = models.BooleanField(
+        null=True, help_text="Acknowledged .gov acceptable use policy"
+    )
+
+    def can_submit(self):
+        """Return True if this instance can be marked as submitted."""
+        if not Website.string_could_be_domain(self.requested_domain):
+            return False
+        return True
```
Comment (SSPJ) [inline src/registrar/models/models.py:257]: Error handling.

1. `requested_domain` can be null, but `Website.string_could_be_domain()` does not know how to deal with that.
2. `requested_domain` is a `Website` object, but `Website.string_could_be_domain()` does not know how to deal with that, either.

Also, nice demonstration of FSM. 🍪

File: src/registrar/tests/test_models.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+from django.db.utils import IntegrityError
+
+from registrar.models import Contact, DomainApplication, User, Website
+
+
+class TestDomainApplication(TestCase):
+    def test_empty_create_fails(self):
+        """Can't create a completely empty domain application."""
+        with self.assertRaisesRegex(IntegrityError, "creator"):
+            DomainApplication.objects.create()
```
Comment (SSPJ) [inline src/registrar/tests/test_models.py:11]: I wouldn't ask you to exhaustively test the FSM, but it'd be nice to have at least one test case to minimally prove that you can't, say, create an application in the `APPROVED ` status.

File: src/registrar/models/models.py [hunk 9]
```diff
@@ -83,3 +87,179 @@ def __str__(self):
...
+
+    @transition(
+        field="status", source=STARTED, target=SUBMITTED, conditions=[can_submit]
```
Comment (SSPJ) [inline src/registrar/models/models.py:260]: General question: what is the difference of `conditions=[can_submit]` vs calling `string_could_be_domain` inside of the body of `submit` and raising if the result is False? Is there a reason to prefer one over the other?

**Review by SSPJ — COMMENTED** [review]
Nice work here, Neil! 💥 

Aside from the inline comments, I think overall that many of the database fields should be indexed, as it will be useful to sort and query by those fields later on.

The Basics
- [x] I read the code
- [x] I ran the code
 
Documentation
- ❌  README (or other documentation) is correct and updated
    You'd mentioned wanting to say something about the FSM in the documentation. Consider including the ADR on this PR, as we've set a precedent of merging the ADRs with the code.
- [x] all new functions and methods are commented
    - [x] using plain language
- ❌  licenses for new dependencies are recorded, if any
- ❓  licenses for updated dependencies have been checked for accuracy, if any
- ❓  new components have been added to the system diagram
 
Security
- N/A security false positives are documented
- N/A data from external sources is cleaned and clearly marked
- N/A significant user actions are logged to a database table
 
Reliability
- ❌  good level of error handling
    - [ ] interactions with external systems are wrapped in try/except
- [x] functionality is tested with unit or integration tests

File: src/registrar/config/settings.py [hunk 10]
```diff
@@ -90,6 +90,7 @@
     "registrar",
     # Our internal API application
     "api",
+    "django_fsm",
```
[the remark this answers is not in the document]
Comment (neilmb) [reply src/registrar/config/settings.py:93]: I had been monkey-patching some of its code which required it to go last, but I took that out of the PR. I'll move it back up above our stuff. (Fascinating (?) tidbit: Django searches INSTALLED_APPS [in reverse](https://docs.djangoproject.com/en/4.1/howto/custom-management-commands/#overriding-commands and commands found later override those found earlier.)

File: src/registrar/models/models.py (hunk 7 above)
Comment (neilmb) [reply src/registrar/models/models.py:257]: Refactored.

File: src/registrar/models/models.py (hunk 9 above)
Comment (neilmb) [reply src/registrar/models/models.py:260]: Leftover from my demonstration. I'm going to recommend raising exceptions inside of `submit` and I'll change it to work that way and include comments to that effect.

File: src/registrar/tests/test_models.py (hunk 8 above)
Comment (neilmb) [reply src/registrar/tests/test_models.py:11]: Added a couple test cases for the `submit` method.

**Review by SSPJ — APPROVED** [review]
LGTM 👍

**Pull request merged**